### ISSUE TITLE
MemoryMappedFile.CreateFromFile constraints on Unix should be loosened to support IFCHR

### ISSUE THREAD

**Issue by stephentoub** [dotnet/corefx#30686]
MemoryMappedFile.CreateFromFile enforces a variety of constraints about the target file.  Those constraints make sense when working with disk-based files, but they don't all make sense when working with special files.  For example, it should be possible to create a map for /dev/mem, but such a special file has zero length, which causes a bunch of the checks in both MemoryMappedFile and MemoryMappedView to fail.

**Comment by kasper3** [comment]
Could there be a separate CreateFromMemory API with an extra overload for `Memory<T>`?

**Comment by wfurt** [comment]
So we would need one more constructor taking length and maybe offset?  We will need that for underlying mmap() and ability to map section regardless of we think about the file size may be useful. 
Error check can always come from underlying system call.

**Comment by stephentoub** [comment]
> We will need that for underlying mmap() and ability to map section regardless of we think about the file size may be useful.

Why?  You can already provide this when creating the MemoryMappedView.  Is there a specific piece that's missing there?

My assumption with regards to this issue is that we don't need any additional surface area, and rather than we can simply remove some of the checks being done when they're not relevant.

**Comment by danmosemsft** [comment]
@carlossanlop seems like a feature request, not required to ship 3.0. If you agree please move to Future.

**Comment by los93sol** [comment]
Would creating a MemoryMappedFile for a socket fall in this scenario as well?  I'm doing some stuff with AF_PACKET and raw sockets and trying to sort out how to leverage mmap with it to avoid the system call to recvmsg for each and every packet and this looks like it would be my answer without dealing with all of it on my own.

**Comment by stephentoub** [comment]
> Would creating a MemoryMappedFile for a socket fall in this scenario as well?

No.  This issue is specifically about CreateFromFile, which takes a file path.

**Comment by karelz** [comment]
Triage: @stephentoub any particular reason to have this one in 5.0? The team does not feel strongly about priority of this. Maybe we are missing something.

**Comment by stephentoub** [comment]
IIRC, this had to do with various IoT scenarios, e.g. from the linked duplicate in https://github.com/dotnet/corefx/issues/32827#issue-370292487, so it would depend on how we want to prioritize those.

**Comment by hopix** [comment]
When developing apps for IoT/embedded Linux there is usually close interaction with peripherals. In lots of cases existing kernel drivers can be used to interact with peripherals. But for custom peripherals or special needs I found it is usually easiest (considering development cost and performance) directly access peripheral controller over special memory mapped file (common example is the /dev/mem).

I vote for this.

**Comment by wfurt** [comment]
I put up PR since this is pretty simple.  Note that `/dev/mem` will not be available unless running as root on most systems. (and my give grief even then if SELinux is enabled) 
It would be great to do some verification with GPIO or some other HW.
